File: S02/week13/sejun/LastStoneWeight.java
// 2ms, 40.67mb

class Solution {
		private Queue<Integer> priorityQueue = new PriorityQueue<>((h1, h2) -> h1 < h2 ? 1 : ((h1.equals(h2)) ? 0 : -1));

		public int lastStoneWeight(int[] stones) {
			for (int stone : stones) {
				priorityQueue.add(stone);
			}

			while (priorityQueue.size() > 1) {
				int result = smash(priorityQueue.poll(), priorityQueue.poll());

				if (result > 0) {
					priorityQueue.add(result);
				}
			}

			if (priorityQueue.isEmpty()) {
				return 0;
			}

			return priorityQueue.poll();
		}

		public int smash(int h1, int h2) {
			return h1 - h2;
		}
	}

File: S02/week13/seokjoon/LastStoneWeight.kt
//154ms, 33.84mb

class Solution {
    fun lastStoneWeight(stones: IntArray): Int {
        val pq = PriorityQueue<Int>(reverseOrder()) // Max Heap

        stones.forEach {
            pq.add(it)
        }

        while(pq.size > 1){
            val y = pq.poll()
            val x = pq.poll()
            if(pq.isEmpty() && x == y ) return 0

            if(x != y){
                pq.add(y-x)
            }
        }
        return pq.poll()

    }
}
